### PR TITLE
[SPARK-38548][SQL] New SQL function: try_sum

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Add a new SQL function: try_sum. It is identical to the function `sum`, except that it returns `NULL` result instead of throwing an exception on integral/decimal value overflow.
Note it is different from sum when ANSI mode is off:
| Function         | Sum                                | TrySum      |
|------------------|------------------------------------|-------------|
| Decimal overflow | Return NULL                        | Return NULL |
| Integer overflow | Return lower 64 bits of the result | Return NULL |

### Why are the changes needed?

* Users can manage to finish queries without interruptions in ANSI mode.
* Users can get NULLs instead of unreasonable results if overflow occurs when ANSI mode is off. For example
```
> SELECT sum(col) FROM VALUES (9223372036854775807L), (1L) AS tab(col);
-9223372036854775808

> SELECT try_sum(col) FROM VALUES (9223372036854775807L), (1L) AS tab(col);
NULL
```

### Does this PR introduce _any_ user-facing change?

Yes, a new SQL function: try_sum which is identical to the function `sum`, except that it returns `NULL` result instead of throwing an exception on integral/decimal value overflow.

### How was this patch tested?

UT

### CODE REVIEW

**Comment by srielau** [comment]
@gengliangwang Why should we propagate the bad overflow behavior for try_sum() even in ANSI_OFF?

**Comment by gengliangwang** [comment]
> @gengliangwang Why should we propagate the bad overflow behavior for try_sum() even in ANSI_OFF?

try_sum() always returns NULL on overflow. It is different from sum().

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 1]
```diff
@@ -452,6 +452,8 @@ object FunctionRegistry {
     expression[TrySubtract]("try_subtract"),
     expression[TryMultiply]("try_multiply"),
     expression[TryElementAt]("try_element_at"),
+    // expressionBuilder("try_sum", TrySumExpressionBuilder),
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:455]: Do we need this comment?

Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: Removed. Thanks!

File: sql/core/src/test/resources/sql-tests/results/ansi/try_aggregates.sql.out [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+0 00:00:02.000000000
+
+
+-- !query
+SELECT try_sum(col) FROM VALUES (interval '106751991 04:00:54.775808' DAY TO SECOND), (interval '1 seconds') AS tab(col)
+-- !query schema
+struct<>
+-- !query output
+java.lang.ArithmeticException
+long overflow
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/ansi/try_aggregates.sql.out:83]: why this doesn't return null?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Sum.scala [hunk 3]
```diff
@@ -110,9 +95,10 @@ case class Sum(
         // in case the input is nullable. The `sum` can only be null if there is no value, as
         // non-decimal type can produce overflowed value under non-ansi mode.
         if (child.nullable) {
```
[the remark this answers is not in the document]
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Sum.scala:97]: It won't work.
The result of the following can be `child.cast(resultType)`
```
coalesce(TryEval(...), child.cast(resultType))
```

File: sql/core/src/test/resources/sql-tests/results/ansi/try_aggregates.sql.out (hunk 2 above)
Comment (gengliangwang) [reply sql/core/src/test/resources/sql-tests/results/ansi/try_aggregates.sql.out]: The left side itself overflow after converting to long type

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Sum.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Sum.scala:97]: Seems `child.cast(resultType)` will never overflow given how we define `resultType`

[the remark this answers is not in the document]
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Sum.scala:97]: yeah, when overflow happens, the result of update becomes `child.cast(resultType)`

**Review by cloud-fan — APPROVED** [review]

**Comment by gengliangwang** [comment]
@MaxGekk I would like to port this to branch-3.3 as well.  Are you OK with it?

**Comment by MaxGekk** [comment]
> @MaxGekk I would like to port this to branch-3.3 as well. Are you OK with it?

Let's wait for the moment when we finish the allow list of changes for Spark 3.3.

**Comment by cloud-fan** [comment]
I think it's in the allowlist? Both @gatorsmile and I mentioned this in the dev list and I don't see any objections.

**Comment by gengliangwang** [comment]
Thanks, merging to master/3.3

**Comment by dongjoon-hyun** [comment]
Thank you, @gengliangwang and all.

**Pull request closed without merging**